Samples/ExampleHost/CountsProjector.cs

```diff
@@ -0,0 +1,336 @@
+                .AsUpdateOf(anEvent => anEvent.DocumentNumber)
+                .Using((document, anEvent) =>
+                {
+                    ValidityPeriod period = document.GetOrAddPeriod(anEvent.Sequence);
+                    period.Status = "Valid";
+
+                    ValidityPeriod lastValidPeriod = document.Periods.LastOrDefault(aPeriod => aPeriod.Status == "Valid");
+
+                    ValidityPeriod[] contiguousPeriods = GetPreviousContiguousValidPeriods(document.Periods, lastValidPeriod)
+                        .OrderBy(x => x.Sequence).ToArray();
+
+                    document.StartDateTime = contiguousPeriods.Any() ? contiguousPeriods.First().From : DateTime.MinValue;
+                    document.EndDateTime = contiguousPeriods.Any() ? contiguousPeriods.Last().To : DateTime.MaxValue;
+                });
+
+            documentMapBuilder
+                .Map<ValidityPeriodClosedEvent>()
+                .AsUpdateOf(anEvent => anEvent.DocumentNumber)
+                .Using((document, anEvent) =>
+                {
+                    ValidityPeriod period = document.GetOrAddPeriod(anEvent.Sequence);
+                    period.Status = "Closed";
+                    period.To = anEvent.ClosedAt;
+                });
+
+            documentMapBuilder
+                .Map<ValidityPeriodCanceledEvent>()
+                .AsUpdateOf(anEvent => anEvent.DocumentNumber)
+                .Using((document, anEvent) =>
+                {
+                    ValidityPeriod period = document.GetOrAddPeriod(anEvent.Sequence);
+                    period.Status = "Canceled";
+                });
+
+            documentProjector = new NHibernateProjector<DocumentCountProjection, string, ProjectorState>(
+                sessionFactory,
+                documentMapBuilder,
+                (projection, identity) => projection.Id = identity,
+                new[] { countryProjector })
+            {
+                BatchSize = 20,
+                Cache = cache,
+                ExceptionHandler = OnException,
+                
+                // Make sure that we no longer process corrupted projections. 
+                Filter = p => !p.Corrupt
+            };
+        }
+
+        private static IEnumerable<ValidityPeriod> GetPreviousContiguousValidPeriods(IEnumerable<ValidityPeriod> allPeriods,
+            ValidityPeriod period)
+        {
+            while (period != null)
+            {
+                yield return period;
+
+                ValidityPeriod previousPeriod =
+                    allPeriods.SingleOrDefault(p => p.Status == "Valid" && p.To.Equals(period.From) && p.Sequence != period.Sequence);
+
+                period = previousPeriod;
+            }
+        }
+
+        private async Task<NHibernate.ExceptionResolution> OnException(ProjectionException exception, int attempts)
+        {
+            if (exception.TransactionBatch.Count > 1 && attempts < 3)

```

Comment:
There are a few improvements that could be applied:
- If the batch consists out of one transaction, then that transaction won't be retried.
- There is no way to determine that we are retrying individual transactions
- The number of attempts should contain every attempt and should be tracked per transaction once we are retrying individually
- If one of the transactions fails during a `RetryIndividual`, then the other transactions should also be retried (also mentioned in earlier comment)

Reply:
> If the batch consists out of one transaction, then that transaction won't be retried.

Completely redesigned the exception handler to take into account whether we're dealing with a transient vs non-transient exception.

> There is no way to determine that we are retrying individual transactions

You can see that from the size of the batch and the number of transactions.

> The number of attempts should contain every attempt and should be tracked per transaction once we are retrying individually

Agreed

> If one of the transactions fails during a RetryIndividual, then the other transactions should also be retried (also mentioned in earlier comment)

Agreed